### ISSUE TITLE
Cannot inject Renderer2 when the application uses Renderer3!

### ISSUE THREAD

**Issue by tk2rush90** [angular/angular#44897]
### Which @angular/* package(s) are the source of the bug?

core, forms

### Is this a regression?

No

### Description

![image](https://user-images.githubusercontent.com/30546713/151698166-bc6d8fe6-5179-47bc-981c-9e91d83e4c69.png)

In my code, I was injecting some providers when calling `createComponent()` method of `ViewContainerRef`. It worked well.
But after upgrading Angaulr version from 13.1 to 13.2, I got 'Cannot inject Renderer2 when the application uses Renderer3!' error .

The code works well if I remove the options for `createComponent()` method, however, I don't want to remove that.
Please help with this issue 😭

### Please provide the exception or error you saw

```true
core.mjs:6468 ERROR Error: Cannot inject Renderer2 when the application uses Renderer3!
    at getOrCreateRenderer2 (core.mjs:21014:1)
    at injectRenderer2 (core.mjs:21025:1)
    at Renderer2.__NG_ELEMENT_ID__ (core.mjs:21009:1)
    at getOrCreateInjectable (core.mjs:3383:1)
    at Module.ɵɵdirectiveInject (core.mjs:14389:1)
    at NodeInjectorFactory.DefaultValueAccessor_Factory [as factory] (forms.mjs:254:1)
    at getNodeInjectable (core.mjs:3548:1)
    at searchTokensOnInjector (core.mjs:3485:1)
    at getOrCreateInjectable (core.mjs:3429:1)
    at ɵɵdirectiveInject (core.mjs:14389:1)
defaultErrorLogger @ core.mjs:6468
handleError @ core.mjs:6515
handleError @ core.mjs:10940
executeListenerWithErrorHandling @ core.mjs:14963
wrapListenerIn_markDirtyAndPreventDefault @ core.mjs:14998
(익명) @ platform-browser.mjs:466
invokeTask @ zone.js:406
onInvokeTask @ core.mjs:25459
invokeTask @ zone.js:405
runTask @ zone.js:178
invokeTask @ zone.js:487
invokeTask @ zone.js:1600
globalZoneAwareCallback @ zone.js:1626
```

### Please provide the environment you discovered this bug in (run `ng version`)

```true
Angular CLI: 13.2.0
Node: 14.17.0
Package Manager: npm 6.14.13
OS: win32 x64

Angular: 13.2.0
... animations, cli, common, compiler, compiler-cli, core, forms
... platform-browser, platform-browser-dynamic, router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1302.0
@angular-devkit/build-angular   13.2.0
@angular-devkit/core            13.2.0
@angular-devkit/schematics      13.2.0
@schematics/angular             13.2.0
rxjs                            7.4.0
typescript                      4.5.5
```

**Comment by AndrewKushnir** [comment]
@tk2rush90 could you please create a repro using [Stackblitz](https://stackblitz.com/), so that we can investigate this issue? Thank you.

**Comment by tk2rush90** [comment]
@AndrewKushnir 

Here is the link:
https://stackblitz.com/edit/github-jc32pa?file=src%2Fapp%2Fsacheon-bus%2Fcomponents%2Fbus-search%2Fbus-search.component.ts

Please check the comment. Thanks.

**Comment by tk2rush90** [comment]
![image](https://user-images.githubusercontent.com/30546713/151728582-38b8b7a5-1b81-49d0-a486-4df53d603016.png)

When calling the `createComponent()` method with `ComponentFactory`, the error doesn't appear.
I think the injected `Renderer2` in `IconComponent` is the cause of this error.

![image](https://user-images.githubusercontent.com/30546713/151729132-135a3d5b-d6ce-467d-adc1-9bb69d19f47a.png)

**Comment by atscott** [comment]
@tk2rush90 The demo link does not run on stackblitz. I also tried downloading, installing dependencies, and running with the CLI and that also didn't work. Can you please provide a working reproduction?

**Comment by alxhub** [comment]
I was able to get the example working and reproduce the problem as described. Here's a trivial reproduction: https://stackblitz.com/edit/ng13-broken-renderer3?file=src/app.ts

It's actually a fascinating bug, and is a combination of several factors:

At the root of the issue is an assumption Angular makes: if the injector being used to create a component doesn't have a binding for `RendererFactory2`, then a factory that creates `Renderer3` (that is, `document`) is used. Exactly this happens for the dynamically instantiated component in the example. That component then attempts to inject `Renderer2`, which causes the crash.

So why is `RendererFactory2` missing? Firstly, the example app creates an injector via `Injector.create([...])`. This injector doesn't inherit from the app injector, so it doesn't have a binding for `RendererFactory2`. This didn't used to be a problem, however, since `ViewContainerRef.createComponent()` tries to create a chained injector using the given injector and the component's `NgModuleRef`. Even if the given injector was missing `RendererFactory2`, the `NgModuleRef` (usually the app injector or derived from it) would have it!

Where did the `NgModuleRef` come from? It's not common to explicitly pass that argument to `ViewContainerRef.createComponent()`. Previously that API required a `ComponentFactory`, though, which _has_ an internal `NgModuleRef` obtained from its `ComponentFactoryResolver`.

In a recent Angular though, we introduced a factory-less API for `createComponent()`, which made it easy to call the API with no `NgModuleRef`. So the _only_ tokens available for injection when creating this component are those in the user-provided injector, which does not include `RendererFactory2`. Therefore, `Renderer3` gets selected, and the component crashes!
